Require group permission to toggle scenery cluster, closes #3045

dlee13 opened the pull request:
#3167

IntelOrca commented:
Anyone could edit window_scenery_is_build_cluster_tool_on to bypass the permission check. It needs to check the permission on the server side.

wolfreak100 commented:
I'm glad i didn't worry about trying to update my version of the code if it was just gonna be brushed off because a _minority_ of the players might know how to memory edit.

Gymnasiast commented:
@wolfreak100 : If you open a PR, you should be able to handle feedback, including being told it isn't up to scratch.

Gymnasiast commented on data/language/english_uk.txt:
```diff
@@ -4015,7 +4016,7 @@ STR_5709    :Rename Group
 STR_5711    :Enter new name for this group:
 STR_5712    :Can't modify permission that you do not have yourself
-STR_5713    :Kick Player
+STR_5713    :Can't remove default group
```
Why did you replace this string?

dlee13 replied:
For my second commit, I picked this string id so that my string could be grouped together with the other strings related to modifying player groups. I added my string id `STR_CANT_REMOVE_DEFAULT_GROUP = 5713,` before I went to the language file and saw that there was already a `STR_5713`. Since string id `5713` was missing before I added mine, that meant it was unused. And there was no reason to reserve it, so I decided to go ahead with it.

ef4fc84 shows that `STR_ACTION_KICK_PLAYER` was moved from `5713` to `5659`.

dlee13 replied:
For my second commit, I picked this string id so that my string could be grouped together with the other strings related to modifying player groups. I added my string id `STR_CANT_REMOVE_DEFAULT_GROUP = 5713,` before I went to the language file and saw that there was already a `STR_5713`. Since string id `5713` was missing before I added mine, that meant it was unused. And there was no reason to reserve it, so I decided to go ahead with it.

ef4fc84 shows that `STR_ACTION_KICK_PLAYER` was moved from `5713` to `5659`.

dlee13 commented:
The server can't tell the difference between when just a single small scenery object is placed and when the cluster tool is used because when a client uses the cluster tool, it just sends the place scenery command several times. There are no flags sent with that command to indicate that the cluster tool was used. And even if I added a flag for the client to send and the server to check for, someone could build their own copy of the game that omits it.

There is not any way the permissions system can be changed to make the server side check useful since someone could make their client send the command regardless of the server saying it's not allowed. (Come to think of it, there's nothing stopping anyone from making their cluster tool send the place scenery command thousands of times. Now that's scary. :fearful: )

Limiting the number of commands that can be run in an amount of time doesn't sound great. The only solution I can see working would be just making the usage of the cluster tool a separate game command. If we go with that, I will remove the permission check for toggling the button. So players would be free to toggle that regardless of permissions, but they would be permission checked if they tried to place scenery with the button toggled on.

To be honest, I actually didn't want this pull request to be accepted as is either, haha. I implemented it as @zsilencer suggested, but it looked hacky to me.

IntelOrca commented:
Yes it would have to be a separate game command or a flag of the existing game command.

dlee13 commented:
Which would you prefer?

IntelOrca commented:
Flag

dlee13 commented:
I feel that the separate game command is the better solution as checking the flag would require putting code specific to `GAME_COMMAND_PLACE_SCENERY` in `Server_Handle_GAMECMD()`. The separate game command is cleaner while the latter implementation is harder to maintain.

Gymnasiast commented:
In the end, we decided to take the other PR and implement the missing parts in that ourselves. Thank you for your contribution.
